pennylane/templates/state_preparations/cosine_window.py

```diff
@@ -0,0 +1,134 @@
+
+    .. note::
+
+        The wave function is shifted by :math:`\frac{\pi}{2}` units so that the window is centered.
+
+
+    **Details:**
+
+    * Number of wires: Any (the operation can act on any number of wires)
+    * Number of parameters: 0
+    * Gradient recipe: None
+
+
+    Args:
+        wires (Sequence[int] or int): the wire(s) the operation acts on
+        id (str): custom label given to an operator instance,
+            can be useful for some applications where the instance has to be identified.
+
+    **Example**
+
+    >>> dev = qml.device('default.qubit', wires=2)
+    >>> @qml.qnode(dev)
+    ... def example_circuit():
+    ...     qml.CosineWindow(wires=range(2))
+    ...     return qml.probs()
+    >>> print(example_circuit())
+    [1.87469973e-33 2.50000000e-01 5.00000000e-01 2.50000000e-01]
+    """
+    num_wires = AnyWires
+    num_params = 0
+    """int: Number of trainable parameters that the operator depends on."""
+
+    def __init__(self, wires, id=None):
+        super().__init__(wires=wires, id=id)
+
+    @staticmethod
+    def compute_decomposition(
+        *args, wires, **kwargs
+    ):  # pylint: disable=arguments-differ,unused-argument
+        r"""Representation of the operator as a product of other operators (static method).
+        It is efficiently decomposed from one QFT over all qubits and one-qubit rotation gates.
+
+        Args:
+            wires (Iterable, Wires): the wire(s) the operation acts on
+
+        Returns:
+            list[Operator]: decomposition into lower level operations
+        """
+
+        decomp_ops = []
+
+        decomp_ops.append(qml.Hadamard(wires=wires[-1]))
+        decomp_ops.append(qml.RZ(np.pi, wires=wires[-1]))
+        decomp_ops.append(qml.adjoint(qml.QFT)(wires=wires))
+
+        for ind, wire in enumerate(wires):
+            decomp_ops.append(qml.PhaseShift(np.pi * 2 ** (-ind - 1), wires=wire))
+
+        return decomp_ops
+
+    def state_vector(self, wire_order=None):
+        num_op_wires = len(self.wires)
+        op_vector_shape = (2,) * num_op_wires
+        vector = np.array(
+            [
+                np.sqrt(2)
+                * np.cos(-np.pi / 2 + np.pi * x / 2 ** (num_op_wires))
+                / np.sqrt(2**num_op_wires)
+                for x in range(2**num_op_wires)
+            ]
+        )
+        op_vector = math.reshape(vector, op_vector_shape)
+
+        if wire_order is None or Wires(wire_order) == self.wires:
+            return op_vector
+
+        wire_order = Wires(wire_order)
+        if not wire_order.contains_wires(self.wires):
+            raise WireError(f"Custom wire_order must contain all {self.name} wires")
+
+        num_total_wires = len(wire_order)
+        indices = tuple(
+            [Ellipsis] + [slice(None)] * num_op_wires + [0] * (num_total_wires - num_op_wires)
+        )
+        ket_shape = [2] * num_total_wires
+
+        ket = np.zeros(ket_shape, dtype=np.complex128)
+        ket[indices] = op_vector
```

Comment:
Can it be rewritten in a better way?

Reply:
I didn't modify this part from [here](https://github.com/PennyLaneAI/pennylane/blame/21ec4d4ed30bde24894f3db5a9a38222ea05db1d/pennylane/ops/qubit/state_preparation.py#L220) in StatePrep. I'll talk to the person who write this to double check 👍

Reply:
Fine if it is consistent with the existing code.

Reply:
Audit told me that in our case we don't have to worry about `wire_order`. This was a particular problem associated with StatePrep. The code is now much cleaner!

Reply:
In the end I had to put `wire_order` but I have simplified the code. In this case `num_op_wires` was not necessary. From my side all changes has been made